### PR TITLE
Improve helper functions and organization

### PR DESCRIPTION
**Pull request by sethrj**
Needed for #176 .

- Refactor the 2D interpolator so that the interpolated x value can be reused
- Move interpolation from `base/` to `physics/grid` which is a better fit
- Fix up minor issues with random distributions

### CODE REVIEW

File: src/physics/grid/TwodSubgridCalculator.i.hh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+CELER_FUNCTION
+TwodSubgridCalculator::TwodSubgridCalculator(const TwodGridData& grids,
+                                             const Values&       storage,
+                                             InterpT             x_loc)
+    : grids_{grids}, storage_(storage), x_loc_(x_loc)
+{
+    CELER_EXPECT(grids);
+    CELER_EXPECT(grids.values.back() < storage.size());
+    CELER_EXPECT(x_loc.index + 1 < grids.x.size());
+    CELER_EXPECT(x_loc.fraction >= 0 && x_loc_.fraction <= 1);
```
Comment (amandalund) [inline src/physics/grid/TwodSubgridCalculator.i.hh:25]: `x_loc_.fraction < 1`?

File: src/physics/grid/TwodGridInterface.hh [hunk 2]
```diff
@@ -32,10 +36,12 @@ struct TwodGridData
...
     {
-        CELER_EXPECT(ix < this->x.size());
-        CELER_EXPECT(iy < this->y.size());
-        return ItemId<real_type>{this->values.front().get()
-                                 + ix * this->y.size() + iy};
+        CELER_EXPECT(ix <= this->x.size());
+        CELER_EXPECT(iy <= this->y.size());
+        size_type index = ix * this->y.size() + iy;
+
+        CELER_ENSURE(index <= this->x.size() * this->y.size());
```
Comment (amandalund) [inline src/physics/grid/TwodGridInterface.hh:43]: Wait why the changes to `<= size`? Shouldn't it be `<`?

File: src/physics/grid/TwodSubgridCalculator.i.hh (hunk 1 above)
Comment (sethrj) [reply src/physics/grid/TwodSubgridCalculator.i.hh:25]: Originally it was `<`, but I figured that interpolation on the last point would be easy to do by constructing with `<=`. Currently though it's always `<` though, so I'll change it to the more restrictive setting and add a note that `==1` should also be OK.

File: src/physics/grid/TwodGridInterface.hh (hunk 2 above)
Comment (sethrj) [reply src/physics/grid/TwodGridInterface.hh:43]: Sometimes it's useful to construct ranges by giving past-the-end values -- but I don't I ended up using it that way, so I'll revert this change too.

**Review by amandalund — APPROVED** [review]
Looks great Seth!

**Comment by sethrj** [comment]
Thanks for the prompt and useful feedback @amandalund !
